### ISSUE TITLE
Leaflet 1.0 compatibility?

### ISSUE THREAD

**Issue by bardu** [perliedman/leaflet-routing-machine#170]
I use Leaflet 1.0 BETA 1 and the latest master, however the following code
```javascript
L.Routing.control({
      waypoints: [
                               L.latLng(coords),
                               L.latLng(_userMarker.getLatLng())
      ],
      instructions: false,
      serviceUrl: 'https://router.project-osrm.org/viaroute'
}).addTo(_map);
```
yields the error:
> TypeError: undefined is not an object (evaluating 't.call')
> whenReadyleaflet.min.js:4:30839
> addLayerleaflet.min.js:5:1556
> onAddleaflet-routing-machine.min.js:2:7369
> addToleaflet.min.js:7:16744
> _findVehicleBtnHandlermap.js:107
> _clickmap.js:153
> hleaflet.min.js:6:25621
> dispatchEvent
> _simulateEventleaflet.min.js:7:11276
> _onUpleaflet.min.js:7:10854
> hleaflet.min.js:6:25621

Any ideas?

**Comment by perliedman** [comment]
@bardu yes, for Leaflet 1.0 compatibility, you need to use the [leaflet-master](https://github.com/perliedman/leaflet-routing-machine/tree/leaflet-master) branch, which has some small changes.

Note that this branch is currently not 100% up to date with LRM's master branch!

**Comment by bardu** [comment]
@perliedman I have used this branch as well but it doesn't make any difference, same error.

**Comment by perliedman** [comment]
@bardu ok, I'll have to investigate!

**Comment by perliedman** [comment]
@bardu ok, I've made a fresh branch with Leaflet 1.0 support: https://github.com/perliedman/leaflet-routing-machine/tree/leaflet-1.0

I've verified that the example works fine with Leaflet 1.0.0-beta.2.

I've also released this as leaflet-routing-machine 3.0.0-beta.1.

**Comment by bardu** [comment]
@perliedman Thanks.
Just wondering what is wrong with my code:
``` javascript
_routingControl = L.Routing.control({
                                            waypoints: [
                                                L.latLng(coords),
                                                L.latLng(_userMarker.getLatLng())
                                            ],
                                            routeLine: function(route) {
                                                var line = L.Routing.line(route, {
                                                        addWaypoints: false
                                                    });
                                                return line;
                                            },
                                            fitSelectedRoutes: true,
                                            show: false,
                                            serviceUrl: 'https://router.project-osrm.org/viaroute'
                                         });

var line = _routingControl.options.routeLine();
```
The last line yields this error:

> TypeError: undefined is not an object (evaluating 'this._route.inputWaypoints')
> _extendToWaypointsleaflet-routing-machine.min.js:2:24548
> initializeleaflet-routing-machine.min.js:2:23962
> eleaflet.min.js:4:2750
> lineleaflet-routing-machine.min.js:2:25394
> routeLinemap.js:112
> (anonymous function)map.js:126
> fireleaflet.min.js:4:4908
> (anonymous function)leaflet-routing-machine.min.js:2:11094
> _routeDoneleaflet-routing-machine.min.js:3:2219
> (anonymous function)leaflet-routing-machine.min.js:3:1111
> (anonymous function)
> eleaflet-routing-machine.min.js:2:881

**Comment by perliedman** [comment]
@bardu ok, you probably shouldn't call `routeLine` like that. The function expects an `IRoute` object as argument, to create the line for. That is, the option is used as a factory for creating lines for routes, not as an accessor to get the current line.

Could you elaborate on what you're trying to do?

**Comment by bardu** [comment]
@perliedman I would like to get the total length of the route prior deciding whether to add the routing control to the map or not.

**Comment by perliedman** [comment]
@bardu ok!

The routing, like much in JavaScript, is done asynchronously, so after the control has been created, the route calculation has not been completed yet, so it's not possible to get the route.

To solve problems like this, events are nice. You can listen for then control's `routesfound` event, which will fire when the route calculation is complete. It will give you an array of alternative routes, and you can check the length of them before deciding if the control should be added.

Check out the API docs for details on how to do this, if you haven't already: http://www.liedman.net/leaflet-routing-machine/api/

**Comment by bardu** [comment]
Got it, thanks.

**Comment by mrohnstock** [comment]
@perliedman leaflet-routing-machine 3.0.0 / 3.0.1 got already released, but support for Leaflet v1.0 wasn't merged. Should we still stick with 3.0.0-beta.1?

**Comment by perliedman** [comment]
Yeah, things got a bit mixed up unfortunately, since I had to make a breaking change to support OSRM 5.

The plan was to make a major refactor and support Leaflet 1.0 for LRM 3.0, but I will have to push those plans forward to 4.0.

I will try to make a 4.0-beta.1 to things a bit clearer, but time for LRM development is unfortunately a bit tight at the moment.
